### PR TITLE
display a link when DONE tweeting

### PR DESCRIPTION
**Pull request by will-weiss**
This shows a message that a tweet is in progress when that is the case and a link to the recent tweet when the tweet is done. We may consider creating a tab on behalf of the user instead of displaying the link, so I didn't spend any energy styling this, instead just focusing on ensuring we have the logic right.

Uses `tweeting` in the `AppState` to represent a single tweet that is either `'NEW' | 'IN_PROGRESS' | 'DONE'` with corresponding `start`, `success`, and `failure` actions, in keeping with the convention we established for fetching the handle.

Only when the tweet is clicked is the feedback cleared, so that at some point we may implement [retries](https://trello.com/c/XJGHUqTV/82-support-retrying-tweeting-on-failure)

Includes a few other quality of life refactors and additional tests.

### CODE REVIEW

File: src/background/responders.ts [hunk 1]
```diff
@@ -78,13 +78,35 @@ export const responders: { [T in Action['type']]: Responder<T> } = {
   CLICK_TAKE_SCREENSHOT(): Partial<AppState> {
     return {}
   },
+  CLICK_TWEET(state): Partial<AppState> {
+    if (state.tweeting?.state !== 'DONE') {
```
Comment (shachar-langer) [inline src/background/responders.ts:82]: Are these assertions for developers or for users? If it's for users, we should probably have some logic that prevents that. If it's for developers, I don't think it's necessary.

File: src/background/subscribe.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ export function subscribe(store: Store<AppState, Action>, browser: typeof window
     previousState = nextState
 
     // Dispatch actions when the extension was clicked
-    if (nextState.popupConnected && !prevState.popupConnected) {
+    if (nextState.popupConnected && !prevState.popupConnected && !nextState.tweeting) {
```
Comment (shachar-langer) [inline src/background/subscribe.ts:18]: Wouldn't this prevent the users to use the extension as long as one tab is tweeting? What happens if the user has a slow connection and it takes time to tweet? Shouldn't we check if we are tweeting from the active tab?

File: src/tests/happy-path.test.ts [hunk 3]
```diff
@@ -12,6 +11,8 @@ import { appendEntity } from '../draft-js-utils'
 
 describe('happy path', () => {
   const mocks = createMocks()
+  const app = () => mocks.popupWindow.document.querySelector('#app-container > .app')!
```
Comment (shachar-langer) [inline src/tests/happy-path.test.ts:14]: What are the advantages of making the `app` a function instead of an element?

File: src/tests/mocks.tsx [hunk 4]
```diff
@@ -48,6 +50,8 @@ export function createMocks(): Mocks {
   Object.assign(global, popupWindowGlobals)
 
   const teardown = () => {
+    // Render a blank div into the app-container before removing globals to trigger any cleanup from the React components themselves
+    ReactDOM.render(<div />, popupWindow.document.getElementById('app-container'))
```
Comment (shachar-langer) [inline src/tests/mocks.tsx:54]: Interesting. I don't understand why we need to do this to trigger a cleanup.

File: src/tests/util.ts [hunk 5]
```diff
@@ -38,5 +40,7 @@ export function whenState<State>(store: Store<State>, predicate: (state: State)
     reject('timeout')
   }, timeoutMillis)
 
+  callback()
```
Comment (shachar-langer) [inline src/tests/util.ts:43]: It's not clear why we need to call the callback

**Comment by shachar-langer** [comment]
As long as I don't open the link, I can't do anything in any other tab

File: src/tests/mocks.tsx (hunk 4 above)
Comment (will-weiss) [reply src/tests/mocks.tsx:54]: emoji-mart had some cleanup steps that rely on there being a global window, so we want those to run before we remove the global window

File: src/tests/happy-path.test.ts (hunk 3 above)
Comment (will-weiss) [reply src/tests/happy-path.test.ts:14]: It can change as the app rerenders

File: src/tests/util.ts (hunk 5 above)
Comment (will-weiss) [reply src/tests/util.ts:43]: The state may already pass the predicate function. So by calling the callback here, we may resolve the promise even if there are no subsequent state changes that trigger store.subscribe

File: src/background/api.ts [hunk 6]
```diff
@@ -18,17 +18,18 @@ function makeTweetRequest(formData: FormData): Promise<Response> {
   })
 }
 
-export const postTweet = async (toBeTweeted: ToBeTweeted, dispatchBackgroundActions: DispatchBackgroundActions) => {
+export const postTweet = async (tab: TabInfo, dispatchBackgroundActions: DispatchBackgroundActions) => {
```
[the remark this answers is not in the document]
Comment (will-weiss) [reply src/background/api.ts:21]: We also use the host, per your PR

File: src/background/actions.ts [hunk 7]
```diff
@@ -32,10 +32,15 @@ export function actions(dispatch: Dispatch<BackgroundAction>): DispatchBackgroun
         payload: { error },
       }),
 
+    postTweetStart: () =>
+      dispatch({
+        type: 'POST_TWEET_START',
+      }),
+
     postTweetSuccess: tweetResult =>
```
[the remark this answers is not in the document]
Comment (will-weiss) [reply src/background/actions.ts:40]: That's part of the payload, I plan on refactoring all these actions away as they're very redundant, stay tuned

File: src/background/responders.ts [hunk 8]
```diff
@@ -78,13 +78,35 @@ export const responders: { [T in Action['type']]: Responder<T> } = {
   CLICK_TAKE_SCREENSHOT(): Partial<AppState> {
     return {}
   },
+  CLICK_TWEET(state): Partial<AppState> {
```
[the remark this answers is not in the document]
Comment (will-weiss) [reply src/background/responders.ts]: 👍 Updated to FOLLOW_TWEET_LINK

File: src/background/responders.ts (hunk 1 above)
Comment (will-weiss) [reply src/background/responders.ts:82]: They're for developers. It's definitely possible to be refactoring and mess this up. Also, by including the assertions in here typescript is able to know that `state.tweeting` has a certain data type in the subsequent lines

File: src/background/responders.ts [hunk 9]
```diff
@@ -199,18 +171,42 @@ export const responders: { [T in Action['type']]: Responder<T> } = {
   SCREENSHOT_CAPTURE_FAILURE(): Partial<AppState> {
     return { alert: 'SCREENSHOT FAILURE' }
   },
-  POST_TWEET_SUCCESS(_state, action): Partial<AppState> {
-    const { url } = action.payload.tweetResult
+  POST_TWEET_SUCCESS(state, action): Partial<AppState> {
+    const tab = state.tabs.get(action.payload.tabId)
+    if (!tab) return {}
```
Comment (shachar-langer) [inline src/background/responders.ts:176]: How can `tab` not exist? and if it doesn't, shouldn't we still clear the feedback?

File: src/background/store.ts [hunk 10]
```diff
@@ -5,8 +5,6 @@ const emptyState: AppState = {
   popupConnected: false,
   focusedWindowId: -1,
   tabs: new Map(),
```
Comment (shachar-langer) [inline src/background/store.ts:7]: what's the initial tweeting value? Shouldn't it be `false`?

File: src/popup/app.tsx [hunk 11]
```diff
@@ -18,13 +18,20 @@ export function App({ state, dispatchUserActions }: AppProps): JSX.Element {
...
-          feedback={activeTab(state)?.feedbackState}
-          user={state.auth.user}
-          pickingEmoji={state.pickingEmoji}
-          dispatchUserActions={dispatchUserActions}
-        />
+        tab && (
+          <div className="app">
+            <Authenticated
+              feedback={tab.feedbackState}
+              host={tab.host!}
```
Comment (shachar-langer) [inline src/popup/app.tsx:27]: I think `host` can be null here

File: src/background/responders.ts (hunk 9 above)
Comment (will-weiss) [reply src/background/responders.ts:176]: Maybe you started tweeting and then closed the tab.

Comment (shachar-langer) [reply src/background/responders.ts:176]: But we're not checking the tab against the current list of tabs. It's the tab we sent to the action. It'll never be null

Comment (shachar-langer) [reply src/background/responders.ts:176]: Nevermind, we're sending the tabId, not the tab. My bad.

**Review by shachar-langer — APPROVED** [review]

**Pull request merged**